### ISSUE TITLE
[GEP-20] Potential race condition for zone-pinning

### ISSUE THREAD

**Issue by timuthy** [gardener/gardener#6905]
**How to categorize this issue?**

/area high-availability
/area control-plane
/kind bug

**What happened**:
The `gardener-resource-manager` adds `podAffinity` and `nodeAffinity` configuration to shoot control-plane pods in order to enforce that all pods are scheduled into the same zone, for `non-HA` or `node` shoots (see https://github.com/gardener/gardener/pull/6579).

We've seen at least one case where this procedure doesn't work reliably:

**Kube Scheduler Logs:**
```
2022-10-25 08:59:14 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/vpa-recommender-7768f7f668-lwq4d\" node=\"z3-85877-pjjbj\" evaluatedNodes=24 feasibleNodes=15","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:59:14 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/kube-scheduler-5b5bc99656-z7zsk\" node=\"z2-59bdb-rjgch\" evaluatedNodes=24 feasibleNodes=15","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:59:14 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/vpa-admission-controller-754774d568-prkx6\" node=\"z3-85877-pjjbj\" evaluatedNodes=24 feasibleNodes=15","pid":"1","severity":"INFO","source":"scheduler.go:621"}
2022-10-25 08:59:13 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/event-logger-5cc489498b-4d9n5\" node=\"z3-85877-pjjbj\" evaluatedNodes=24 feasibleNodes=15","pid":"1","severity":"INFO","source":"scheduler.go:621"}
2022-10-25 08:59:03 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/gardener-resource-manager-d5bb747d-6c2hk\" node=\"z2-59bdb-rjgch\" evaluatedNodes=24 feasibleNodes=15","pid":"1","severity":"INFO","source":"scheduler.go:621"}
2022-10-25 08:58:58 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/kube-controller-manager-795f6dfd94-v9c6j\" node=\"z2-59bdb-rjgch\" evaluatedNodes=24 feasibleNodes=14","pid":"1","severity":"INFO","source":"scheduler.go:621"}
2022-10-25 08:58:57 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/gardener-resource-manager-6d9fb47b95-dsdv8\" node=\"z3-85877-pjjbj\" evaluatedNodes=24 feasibleNodes=15","pid":"1","severity":"INFO","source":"scheduler.go:621"}2022-10-25 08:58:57 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/gardener-resource-manager-6d9fb47b95-mmnbk\" node=\"z2-59bdb-ql77c\" evaluatedNodes=24 feasibleNodes=15","pid":"1","severity":"INFO","source":"scheduler.go:621"}
2022-10-25 08:58:57 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/gardener-resource-manager-6d9fb47b95-wxznf\" node=\"z2-59bdb-rjgch\" evaluatedNodes=24 feasibleNodes=15","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:58:01 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/kube-apiserver-58d6557c4c-tgf9p\" node=\"z3-85877-bf5p6\" evaluatedNodes=24 feasibleNodes=7","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:57:43 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/csi-driver-controller-file-54cb59f56d-gnlbp\" node=\"z3-85877-bf5p6\" evaluatedNodes=24 feasibleNodes=14","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:57:43 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/csi-driver-controller-disk-fc55ffd89-t65dd\" node=\"z3-85877-bf5p6\" evaluatedNodes=24 feasibleNodes=14","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:57:43 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/csi-snapshot-controller-64859bf75c-nn7t7\" node=\"z3-85877-bf5p6\" evaluatedNodes=24 feasibleNodes=15","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:57:42 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/csi-snapshot-validation-86b97cc955-gzh6v\" node=\"z3-85877-bf5p6\" evaluatedNodes=24 feasibleNodes=15","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:57:42 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/cloud-controller-manager-5c988c7599-2t78g\" node=\"z2-59bdb-rjgch\" evaluatedNodes=24 feasibleNodes=15","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:57:10 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/etcd-main-0\" node=\"etcd-z2-654cb-nqfcl\" evaluatedNodes=24 feasibleNodes=9","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:57:10 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/etcd-events-0\" node=\"z2-59bdb-ql77c\" evaluatedNodes=24 feasibleNodes=13","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:56:53 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/bar.infra.tf-apply-p7hjk\" node=\"z3-85877-bf5p6\" evaluatedNodes=24 feasibleNodes=14","pid":"1","severity":"INFO","source":"scheduler.go:621"}	
2022-10-25 08:56:11 {"log":"\"Successfully bound pod to node\" pod=\"shoot--foo--bar/vpn-seed-server-5cd8596bff-zhlzc\" node=\"z2-59bdb-ql77c\" evaluatedNodes=24 feasibleNodes=14","pid":"1","severity":"INFO","source":"scheduler.go:621"}
```

**Gardenlet Logs:**
```
2022-10-25 08:57:55 {"log":{"flow":"Shoot cluster reconciliation","level":"info","logger":"controller.shoot.shoot","msg":"Succeeded","name":"bar","namespace":"garden-foo","operation":"reconcile","task":"Adding zone information to Shoot namespace","ts":"2022-10-25T08:57:55.379Z"}}
```

The logs shows that first pod `shoot--foo--bar/vpn-seed-server-5cd8596bff-zhlzc` is scheduled to a node in zone `z2` but then pod `bar.infra.tf-apply-p7hjk` is scheduled to `z3`, whereas some further pods land again in `z2` but also in `z3`.

It's not yet clear why we see this zone mixture if the following affinity is defined at all times:

```
spec:
  affinity:
    podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector: {}
        topologyKey: topology.kubernetes.io/zone
```

**What you expected to happen**:
All control-plane pods being scheduled into the same zone.

**How to reproduce it (as minimally and precisely as possible)**:
Not clear yet.

**Environment**:

- Gardener version: `v1.58.0`
- Kubernetes version (use `kubectl version`): `v1.23.9`

**Comment by timuthy** [comment]
Update: It's still not clear why and under which circumstances this issue happened. The logs didn't help either to narrow down the issue.
However, it's planned to expand the webhook to handle HA concerns in general and with this endeavor we can improve logging as well. Also, the affinity configuration will slightly change, i.e. in the future Gardenlet will pick zone(s) before any workload is scheduled. This change will allow us to drop the `podAffinity` that can also be the root cause, although that's questionable at the moment.

TL;DR: We plan to adjust the affinity configuration and improve the logging together with @rfranzke. After that we'll carefully observe if the issue happens again. Let's keep this issue in the meantime to post any related updates.
